New decomposition of `QubitUnitary` based on Cartan decomposition (#7474)
PennyLaneAI/pennylane: doc/releases/changelog-dev.md

<h3>Breaking changes 💔</h3>

* Decompositions of two-qubit unitaries via `two_qubit_decomposition` now use 3 CNOTs even if
  just 2 CNOTs would suffice theoretically. This is the price for guaranteeing correctness in
  all edge cases, and will be improved in a later contribution.

>   [(#7474)](https://github.com/PennyLaneAI/pennylane/pull/7474)

Reviewer: I think it's worth erring on the side of breaking change here, since for some users the explicit decomposition _is_ important, and if that changed they may consider it a breaking change.

So I would suggest keeping it here for visibility, but emphasizing:

- Yes decomposition will change, this may/may not affect you depending on your workflow
- However, this does fix a bug which resulted in incorrect results previously
- Overall, numerical results should stay the same.

>   [(#7474)](https://github.com/PennyLaneAI/pennylane/pull/7474)

Author: I would be a bit annoyed if my gate counts grew silently and I could not find out in the changelog why that might be. If a user's workflow did not hit any of the edge cases but was using the two-CNOT scenario, this PR makes things strictly worse for them, no?

<h3>Breaking changes 💔</h3>

> * Decompositions of two-qubit unitaries via `two_qubit_decomposition` now use 3 CNOTs even if

Reviewer 2: Is `two_qubit_decomposition` a user-facing function? 🤔

Reviewer 3: It is public and visible in the documentation, not sure if anyone actually uses it though.

Author: The [compilation hub entry on two-qubit synthesis](https://pennylane.ai/compilation/two-qubit-synthesis) advertises it, for example.

>   just 2 CNOTs would suffice theoretically. This may or may not affect a user's workflow, depending
>   on the numerical matrices provided to the decomposition. This change fixes a previous bug that
>   affected the correctness of the decomposition in some edge cases.

Reviewer 2: I'm still a little confused here 😵. If 2 CNOTs suffice, then why was there a bug? Perhaps this is a better wording (less words 😅).

```suggestion
  just 2 CNOTs would suffice theoretically. This fixes a bug in the implementation for this decomposition
  that affected its correctness in some edge cases.
```

Author: scratch that, given your other feedback I rewrote it.

<h3>Bug fixes 🐛</h3>

> * Implemented a new decomposition of two-qubit `QubitUnitary` operators in `two_qubit_decomposition`
>   based on a type-AI Cartan decomposition. It fixes previously faulty edge cases for unitaries
>   that require 2 or 3 CNOT gates. Now, 3 CNOTs are used for both cases, using one more
>   CNOT than theoretically required in the former case.
>   [(#7474)](https://github.com/PennyLaneAI/pennylane/pull/7474)

Reviewer 2: ```suggestion
* A new decomposition was Implemented for two-qubit `QubitUnitary` operators in `two_qubit_decomposition`
  based on a type-AI Cartan decomposition. It fixes previously faulty edge cases for unitaries
  that require 2 or 3 CNOT gates. Now, 3 CNOTs are used for both cases, using one more
  CNOT than theoretically required in the former case.
  [(#7474)](https://github.com/PennyLaneAI/pennylane/pull/7474)
```

Couple comments on this outside of the suggestion:
- Is there a link we can provide for type-AI Cartan decomposition?
- I found this bug-fix entry to be a lot clearer than the entry in breaking changes 🤔.
